### PR TITLE
detect and report local graph changes

### PR DESCRIPTION
**Pull request by wjwwood**
Fixes https://github.com/ros2/rmw_connext/issues/168

This required, apparently for the first time, that the node is given to the `rmw_destroy_client` and `rmw_destroy_service` calls, so that will change the rmw API.

I'm still debugging some issues with the tests.

### CODE REVIEW

**Comment by wjwwood** [comment]
So, it seems that Connext's request-reply uses some strange topic names. For example, the service server creates:

```
+S add_two_intsRequest <example_interfaces::srv::dds_::AddTwoInts_Request_>
+P add_two_intsReply <example_interfaces::srv::dds_::AddTwoInts_Response_>
```

But the client creates:

```
+S add_two_intsReply_ac1702d4000020e70000000180002003 <example_interfaces::srv::dds_::AddTwoInts_Response_>
+P add_two_intsRequest <example_interfaces::srv::dds_::AddTwoInts_Request_>
```

Note what looks like a GUID attached to the reply topic. It's not clear to me how the replier (service server) delivers the reply to the requester (service client).

I spent the better part of last night trying to figure out a better way to figure out when the service server is available for how Connext implements Request-Reply. I have some ideas, but as of right now this is not ready for merging.

File: rmw_connext_cpp/src/functions.cpp [hunk 1]
```diff
@@ -176,7 +198,7 @@ rmw_create_publisher(
     return nullptr;
   }
 
-  auto node_info = static_cast<ConnextNodeInfo *>(node->data);
+  ConnextNodeInfo * node_info = static_cast<ConnextNodeInfo *>(node->data);
```
Comment (wjwwood) [inline rmw_connext_cpp/src/functions.cpp]: I think I did that so my autocomplete would work, but I can change them back: 90d4c86

**Comment by dirk-thomas** [comment]
This recent change broke `rmw_connext_dynamic_cpp`: http://ci.ros2.org/job/ci_linux/1870/testReport/

While not being actively tested at the moment maybe it can be easily fixed @wjwwood ?

**Comment by wjwwood** [comment]
I believe it is now failing because the wait_for_service is still not implemented, see:

http://ci.ros2.org/job/ci_linux/1870/testReport/(root)/projectroot/requester_replier_cpp__Primitives__rmw_connext_dynamic_cpp/

The tests could be disabled, but I don't know if it is worth it since we're not regularly trying to figure out which failures are known and which are not for that rmw implementation. The fact that this function is not implemented is already tracked in an issue: https://github.com/ros2/rmw_connext/issues/203

I'm not sure about the other failures.

**Comment by dirk-thomas** [comment]
Since the merge numerous very simple tests which only do pub/sub are failing. I don't think the referenced ticket about `wait_for_service` is related to all of them. When I tried it with the new composition demos the executable crashed when calling one of the new functions around `add_information`. Therefore I am pretty sure that this is a regression of this PR.

**Comment by wjwwood** [comment]
Well I'm sorry it is broken, because it wasn't my intention to break it, in fact I spent some extra time trying to keep it up to speed. However, fixing this is not at the top of my priority list because it's not  one of our supported rmw implementations right now.

**Pull request merged**